overlord/hookstate/ctlcmd/kmod.go

```diff
@@ -0,0 +1,197 @@
+		return fmt.Errorf("cannot load module %q, required interface not connected", k.Positional.Module)
+	}
+
+	if err := kmodLoadModule(k.Positional.Module, k.Positional.Options); err != nil {
+		return fmt.Errorf("cannot load module %q: %v", k.Positional.Module, err)
+	}
+
+	return nil
+}
+
+type KModRemoveCmd struct {
+	kmodNopExecuteMixin
+	Positional struct {
+		Module string `positional-arg-name:"<module>" required:"yes" description:"kernel module name"`
+	} `positional-args:"yes" required:"yes"`
+	kmod *kmodCommand
+}
+
+func (k *KModRemoveCmd) Execute([]string) error {
+	context, err := k.kmod.ensureContext()
+	if err != nil {
+		return err
+	}
+
+	attributes, err := kmodFindConnection(context, k.Positional.Module, []string{})
+	if err != nil {
+		return fmt.Errorf("cannot unload module %q: %v", k.Positional.Module, err)
+	}
+
+	if len(attributes) == 0 {
+		return fmt.Errorf("cannot unload module %q, required interface not connected", k.Positional.Module)
+	}
+
+	if err := kmodUnloadModule(k.Positional.Module); err != nil {
+		return fmt.Errorf("cannot unload module %q: %v", k.Positional.Module, err)
+	}
+
+	return nil
+}
+
+// kmodMatchConnection checks whether the given kmod connection attributes give
+// the snap permission to execute the kmod command
+func kmodMatchConnection(attributes map[string]interface{}, moduleName string, moduleOptions []string) bool {
+	load, found := attributes["load"]
+	if !found || load.(string) != "dynamic" {
+		return false
+	}
+
+	if moduleName != attributes["name"].(string) {
+		return false
+	}
+
+	if len(moduleOptions) > 0 {
+		// snapctl can be invoked with options only if the "options" attribute
+		// on the plug is set to "*"
+		optionsAttr, ok := attributes["options"]
+		if !ok || optionsAttr.(string) != "*" {
+			return false
+		}
+	}
+
+	return true
+}
+
+// kmodFindConnections walks through the established connections to find one which
+// is compatible with a kmod operation on the given moduleName and
+// moduleOptions. If found, it returns the connection's attributes.
+var kmodFindConnection = func(context *hookstate.Context, moduleName string, moduleOptions []string) (attributes map[string]interface{}, err error) {
+	snapName := context.InstanceName()
+
+	st := context.State()
+	st.Lock()
+	defer st.Unlock()
+
+	conns, err := ifacestate.ConnectionStates(st)
+	if err != nil {
+		return nil, fmt.Errorf("internal error: cannot get connections: %s", err)
+	}
+
+	for connId, connState := range conns {
+		if connState.Interface != "kernel-module-load" {
+			continue
+		}
+
+		if !connState.Active() {
+			continue
+		}
+
+		connRef, err := interfaces.ParseConnRef(connId)
+		if err != nil {
+			return nil, err
+		}
+
+		if connRef.PlugRef.Snap != snapName {
+			continue
+		}
+
+		modules, ok := connState.StaticPlugAttrs["modules"].([]interface{})
+		if !ok {
+			continue
+		}
+
+		for _, moduleAttributes := range modules {
+			attributes := moduleAttributes.(map[string]interface{})
```

Comment:
Should we do a type check for `map[string]string` since we assume that in `kmodMatchConnection`?